### PR TITLE
copr: support collation aware min/max aggregation functions

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/17123 

Problem Summary: support collation aware min/max aggregation functions

### What is changed and how it works?

Resolve https://github.com/pingcap/tidb/issues/17123.
Support collation aware min/max aggregation functions.

### Check List 

Tests 

- Unit test

### Release note 

- copr: support collation aware min/max aggregation functions

### CODE REVIEW

File: components/tidb_query_vec_aggr/src/impl_max_min.rs [hunk 1]
```diff
@@ -281,6 +387,49 @@ mod tests {
...
+            (Collation::Utf8Mb4GeneralCi, false, vec!["a", "A"], "a"),
+            (Collation::Utf8Mb4BinNoPadding, false, vec!["a", "A"], "A"),
+        ];
+        for (coll, is_max, args, expected) in cases {
+            let function = match_template_collator! {
+                TT, match coll {
+                    Collation::TT => {
+                        match is_max {
+                            true => Box::new(AggFnExtremum4Bytes::<TT, Max>::new()) as Box<dyn AggrFunction>,
+                            false => Box::new(AggFnExtremum4Bytes::<TT, Min>::new()) as Box<dyn AggrFunction>
```
Comment (zhongzc) [inline components/tidb_query_vec_aggr/src/impl_max_min.rs:409]: Silent `make clippy` seems not working well but `PROST=1 make clippy` complains that
```
[2020-06-11T12:09:11.818Z] 407 | /                         match is_max {
[2020-06-11T12:09:11.818Z] 408 | |                             true => Box::new(AggFnExtremum4Bytes::<TT, Max>::new()) as Box<dyn AggrFunction>,
[2020-06-11T12:09:11.818Z] 409 | |                             false => Box::new(AggFnExtremum4Bytes::<TT, Min>::new()) as Box<dyn AggrFunction>
[2020-06-11T12:09:11.818Z] 410 | |                         }
[2020-06-11T12:09:11.818Z]     | |_________________________^ help: consider using an `if`/`else` expression: `if is_max { Box::new(AggFnExtremum4Bytes::<TT, Max>::new()) as Box<dyn AggrFunction> } else { Box::new(AggFnExtremum4Bytes::<TT, Min>::new()) as Box<dyn AggrFunction> }`
```

File: components/tidb_query_vec_aggr/src/impl_max_min.rs [hunk 2]
```diff
@@ -64,6 +65,7 @@ impl<T: Extremum> super::AggrDefinitionParser for AggrFnDefinitionParserExtremum
 
         let out_ft = aggr_def.take_field_type();
         let out_et = box_try!(EvalType::try_from(out_ft.as_accessor().tp()));
+        let out_coll = out_ft.as_accessor().collation();
```
Comment (breeswish) [inline components/tidb_query_vec_aggr/src/impl_max_min.rs:68]: Maybe we can leave the error case here and don't need to manually check for it:

```rust
let out_coll = out_ft.as_accessor().collation()?;
```

or if the above not work:

```rust
let out_coll = box_try!(out_ft.as_accessor().collation());
```

File: components/tidb_query_vec_aggr/src/impl_max_min.rs [hunk 3]
```diff
@@ -88,6 +103,97 @@ impl<T: Extremum> super::AggrDefinitionParser for AggrFnDefinitionParserExtremum
...
+#[derive(Debug, AggrFunction)]
+#[aggr_function(state = AggFnStateExtremum4Bytes::<T>::new(self.ord))]
+pub struct AggFnExtremum4Bytes<T, E>
+where
+    T: Collator,
+    E: Extremum,
+    VectorValue: VectorValueExt<Bytes>,
+{
+    _phantom: std::marker::PhantomData<(T, E)>,
+    ord: Ordering,
```
Comment (breeswish) [inline components/tidb_query_vec_aggr/src/impl_max_min.rs:115]: you can remove ord just like https://github.com/tikv/tikv/pull/8072

File: components/tidb_query_vec_aggr/src/impl_max_min.rs [hunk 4]
```diff
@@ -88,6 +103,97 @@ impl<T: Extremum> super::AggrDefinitionParser for AggrFnDefinitionParserExtremum
     }
 }
 
+#[derive(Debug, AggrFunction)]
+#[aggr_function(state = AggFnStateExtremum4Bytes::<T>::new(self.ord))]
+pub struct AggFnExtremum4Bytes<T, E>
+where
+    T: Collator,
```
Comment (breeswish) [inline components/tidb_query_vec_aggr/src/impl_max_min.rs:110]: Prefer to use `C: Collator`

File: components/tidb_query_vec_aggr/src/impl_max_min.rs [hunk 5]
```diff
@@ -80,6 +82,19 @@ impl<T: Extremum> super::AggrDefinitionParser for AggrFnDefinitionParserExtremum
             src_schema.len(),
         )?);
 
+        if out_et == EvalType::Bytes {
+            return match out_coll {
+                Ok(coll) => {
+                    match_template_collator! {
+                        TT, match coll {
```
Comment (breeswish) [inline components/tidb_query_vec_aggr/src/impl_max_min.rs:89]: Prefer `TT` to be `C`

File: components/tidb_query_vec_aggr/src/impl_max_min.rs [hunk 6]
```diff
@@ -88,6 +103,97 @@ impl<T: Extremum> super::AggrDefinitionParser for AggrFnDefinitionParserExtremum
...
+    }
+}
+
+#[derive(Debug)]
+pub struct AggFnStateExtremum4Bytes<T>
+where
+    VectorValue: VectorValueExt<Bytes>,
+    T: Collator,
+{
+    ord: Ordering,
```
Comment (breeswish) [inline components/tidb_query_vec_aggr/src/impl_max_min.rs:138]: ditto. ord can be removed like #8072

File: components/tidb_query_vec_aggr/src/impl_max_min.rs [hunk 7]
```diff
@@ -88,6 +103,97 @@ impl<T: Extremum> super::AggrDefinitionParser for AggrFnDefinitionParserExtremum
...
+        if value.is_none() {
+            return Ok(());
+        }
+
+        if self.extremum.is_none() {
+            self.extremum = value.clone();
+            return Ok(());
+        }
+
+        match T::sort_compare(&self.extremum.as_ref().unwrap(), &value.as_ref().unwrap()) {
```
Comment (breeswish) [inline components/tidb_query_vec_aggr/src/impl_max_min.rs:179]: maybe consider using `?` operator or `box_try!` if `?` operator doesn't work

File: components/tidb_query_vec_aggr/src/impl_max_min.rs [hunk 8]
```diff
@@ -281,6 +387,49 @@ mod tests {
...
+            (Collation::Binary, false, vec!["a", "A"], "A"),
+            (Collation::Utf8Mb4Bin, false, vec!["a", "A"], "A"),
+            (Collation::Utf8Mb4GeneralCi, false, vec!["a", "A"], "a"),
+            (Collation::Utf8Mb4BinNoPadding, false, vec!["a", "A"], "A"),
+        ];
+        for (coll, is_max, args, expected) in cases {
+            let function = match_template_collator! {
+                TT, match coll {
+                    Collation::TT => {
+                        match is_max {
```
Comment (breeswish) [inline components/tidb_query_vec_aggr/src/impl_max_min.rs:407]: `if is_max { ... } else { ... }`

**Comment by breeswish** [comment]
Also you need to sign-off **all** commits to pass the DCO check. (You will need a rebase).

**Comment by qw4990** [comment]
Thanks for your comments, @breeswish @zhongzc , all comments have been addressed now.

**Comment by breeswish** [comment]
Also do we need to add some integration tests to cover the case?

**Comment by qw4990** [comment]
@sticnarf Sorry for my late reply. All comments have been addressed, PTAL.

**Pull request merged**